### MENU
1. sys  
2. OS  
3. random模块  
4. time模块  
5. json模块  
6. 正则表达式(Regular Expression)
7. 容器相关类 --> set和frozenset  
8. 容器相关类 --> 双端队列deque
9. 容器相关类 --> ChainMap对象
10. 容器相关类 --> OrderedDict对象  
11. 函数相关模块 --> itertools  
12. 函数相关模块 --> functools  
13. decimal

### sys: 代表了Python解释器，主要用于获取和Python解释器相关的信息

In [1]:
print(sys.platform, end='\n\n')
print(sys.version)

win32

3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]


In [2]:
# 获取运行参数
for arg in sys.argv:
    print(arg)

ipykernel_launcher
--ip=127.0.0.1
--stdin=9028
--control=9026
--hb=9025
--Session.signature_scheme="hmac-sha256"
--Session.key=b"24e8815c-4eb9-4549-91e1-ff9ac6e22c86"
--shell=9027
--transport="tcp"
--iopub=9029
--f=C:\Users\Leon\AppData\Local\Temp\tmp-143404vbD708b1409.json


### OS: 主要用于获取程序运行所在操作系统的相关信息

In [3]:
print(os.cpu_count())
print(os.sep)
print(os.pathsep)

8
\
;


### random模块

In [4]:
import random

print(random.__all__)

['Random', 'seed', 'random', 'uniform', 'randint', 'choice', 'sample', 'randrange', 'shuffle', 'normalvariate', 'lognormvariate', 'expovariate', 'vonmisesvariate', 'gammavariate', 'triangular', 'gauss', 'betavariate', 'paretovariate', 'weibullvariate', 'getstate', 'setstate', 'getrandbits', 'choices', 'SystemRandom']


### time模块

In [5]:
import time

# 将当前事件转换为时间字符串
print(time.asctime())

# 将当前时间转换为指定格式的字符串
print(time.strftime('%Y-%m-%d %H:%M:%S'))

# 将指定时间字符串恢复成struct_time对象
print(time.strptime('2020年10月26日', '%Y年%m月%d日'))

Sat May 29 13:54:14 2021
2021-05-29 13:54:14
time.struct_time(tm_year=2020, tm_mon=10, tm_mday=26, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=0, tm_yday=300, tm_isdst=-1)


### json模块

In [6]:
import json

print(json.__all__)

['dump', 'dumps', 'load', 'loads', 'JSONDecoder', 'JSONDecodeError', 'JSONEncoder']


In [7]:
# 将Python对象转换为JSON字符串(元组会被当成数组)
s = json.dumps(['kk', {'favorite': ('coding', None, 'game', 25)}])
s

'["kk", {"favorite": ["coding", null, "game", 25]}]'

In [8]:
# 将简单的字符串准换为JSON字符串
s1 = json.dumps("\"foo\bar")
s1

'"\\"foo\\bar"'

In [9]:
# 将Python的dict对象转换为JSON字符串，并对key排序
s2 = json.dumps({'c': 0, 'b': 0, 'a': 0}, sort_keys=True)
s2

'{"a": 0, "b": 0, "c": 0}'

In [10]:
# 将Python转换为JSON字符串并指定JSON分隔符：在逗号和冒号之后没有空格(默认有空格)
s3 = json.dumps([1, 2, 3, {'x': 5, 'y': 7}], separators=(',', ':'))
s3

'[1,2,3,{"x":5,"y":7}]'

In [11]:
# 指定indent为4，意味着转换的JSON字符串有缩进
s4 = json.dumps({'java': 4, 'python': 6}, sort_keys=True, indent=4)
s4

'{\n    "java": 4,\n    "python": 6\n}'

In [12]:
# 使用JSONEncoder的encode方法将Python对象转换为JSON字符串
s5 = json.JSONEncoder().encode({'names': ('kk', 'master')})
s5

'{"names": ["kk", "master"]}'

将转换的JSON字符串输出到文件中: json.dump(\['Swift', {'Python': 'excellent'}\], f). dumps与dump函数的功能、所支持的基本选项基本相同，只是dumps函数直接返回转换得到的JSON字符串，dump将转换得到的JSON字符串输出到文件中。

In [13]:
# 将JSON字符串恢复成Python列表
result1 = json.loads('["kk", {"favorite": ["coding", null, "game", 25]}]')
result1

['kk', {'favorite': ['coding', None, 'game', 25]}]

In [14]:
# 将JSON字符串恢复成Python字符串
result2 = json.loads('"\\"foo\\"bar"')
result2

'"foo"bar'

In [15]:
# 从文件流恢复JSON列表
# f = open('test.json')
# result3.load(f)

### 正则表达式(Regular Expression)
用于描述一种字符串匹配的模式，它可以用于检查一个字符串是否包含有某个字串，也可用于从字符串中提取匹配的子串，或者对字符串中匹配的子串执行替换操作。

In [16]:
import re

print(re.__all__)

['match', 'fullmatch', 'search', 'sub', 'subn', 'split', 'findall', 'finditer', 'compile', 'purge', 'template', 'escape', 'error', 'Pattern', 'Match', 'A', 'I', 'L', 'M', 'S', 'X', 'U', 'ASCII', 'IGNORECASE', 'LOCALE', 'MULTILINE', 'DOTALL', 'VERBOSE', 'UNICODE']


re.compile(pattern, flags=0): 该函数用于将正则表达式字符串编译成\_sre.SRE\_Pattern对象，该对象代表了正则表达式编译之后在内存中的对象，它可以缓存并复用正则表达式字符串，如果程序需要多次使用同一个正则表达式字符串，可以考虑先编译它:

In [17]:
p = re.compile('abc')
p

re.compile(r'abc', re.UNICODE)

In [18]:
# 调用_sre.SRE_Pattern对象的search()方法
p.search("www.abc.com")

<re.Match object; span=(4, 7), match='abc'>

In [19]:
# 直接使用正则表达式匹配目标字符串
re.search('abc', 'www.abc.com')

<re.Match object; span=(4, 7), match='abc'>

1、re.search(pattern, string, flags=0): 扫描整个字符串，并返字符串中第一处匹配pattern的匹配对象，该函数返回sre.SRE_Pattern对象; 

2、re.match(pattern, string, flags=0): 尝试从字符串的开始位置来匹配正则表达式，如果从开始位置不成功，返回None。该函数返回sre.SRE_Pattern对象，该对象包含的span(n)方法用于获取第n+1个组的匹配位置，group(n)方法用于获取第n+1个组所匹配的子串。

3、search和match的区别在于：match必须从字符串开始处就匹配，但search则可以搜索整个字符串。

In [20]:
re.findall('abc', 'www.abc.com, abc_website', re.I)

['abc', 'abc']

In [21]:
it = re.finditer('abc', 'www.abc.com, abc_website', re.I)

for e in it:
    print(str(e.span()) + " --> " + e.group())

(4, 7) --> abc
(13, 16) --> abc


1、re.fullmatch(pattern, string, flags=0): 该函数要求整个字符串能匹配pattern，如果匹配则返回包含匹配信息的\_sre.SRE_Match对象，否则返回None;  

2、re.sub(pattern, repl, string, count=0, flags=0): 该函数用于将string字符串中所有匹配pattern的内容替换成repl；repl既可以是被替换的字符串，也可以是一个函数。count参数控制最多替换多少次，如果指定count为0，则表示全部替换。

In [22]:
test_date = '2020-10-29'
re.sub(r'-', '/', test_date)

'2020/10/29'

In [23]:
re.sub(r'-', '/', test_date, count=1)

'2020/10-29'

re.split(pattern, string, maxsplit=0, flags=0): 使用pattern对string进行分割，该函数返回分割得到的多个子串组成的列表，maxsplit参数控制最多分割几次：

In [24]:
re.split(',', 'kk, master, monster')

['kk', ' master', ' monster']

In [25]:
re.split(',', 'kk, master, monster', maxsplit=1)

['kk', ' master, monster']

1、re.purge(): 清除正则表达式缓存。  

2、re.escape(pattern): 对模式中除ASCII字符、数值、下划线之外的其他字符进行转义。

In [26]:
re.escape('kk.sun beats monsters to protect the master!')

'kk\\.sun\\ beats\\ monsters\\ to\\ protect\\ the\\ master!'

用圆括号将多个表达式括起来形成组:

In [27]:
m = re.search(r'(abc).(com)', r"www.abc.com.cn")
m

<re.Match object; span=(4, 11), match='abc.com'>

In [28]:
m.group(0)

'abc.com'

In [29]:
# 底层调用的是m.__getitem__(0)
m[0]

'abc.com'

In [30]:
m.span(0)

(4, 11)

In [31]:
m.groups()

('abc', 'com')

上面程序中使用了一个包含两个组的正则表达式，因此程序可以依次获取group(0), group(1), group(2), 即依次获取整个正则表达式所匹配的子串、第一个组所匹配的子串和第二个组所匹配的子串；程序也可以依次获取span(0), span(1) 和 span(2)的值，即依次获取整个正则表达式、第一个组、第二个组所匹配的子串的开始位置和结束位置。如果在正则表达式中为组指定了名字(用 ?P<名字> 为正则表达式的组指定名字)，就可以调用groupdict()方法来获取所有组所匹配的字符串组成的字典，其中作为字典的key:

In [32]:
m2 = re.search(r'(?P<Prefix>abc).(?P<Suffix>com)', r"www.abc.com.cn is a website.")
m2.groupdict()

{'Prefix': 'abc', 'Suffix': 'com'}

### 正则表达式旗标
1、 re.A或re.ASCII: 改旗标控制\w, \W, \b, \B, \d, \D, \s, \S 只匹配ASCII字符，而不匹配所有的Unicode字符。

2、 re.I或re.IGNORECASE: 使用正则表达式匹配时不区分大小写。

3、 re.L或re.LOCALE: 根据当前区域设置使用正则表达式匹配时不区分大小写，该旗标只能对bytes模式起作用。

4、 re.M或re.MULTILINE: 多行模式的旗标，当指定该旗标后，'^'能匹配字符串的开头和每行的开头; '$'能匹配字符串的末尾和每行的末尾。

5、 re.S或re.DOTALL: 让(.)能匹配包括换行符在内的所有字符，如果不指定该旗标，则(.)只能匹配不包括换行符的所有字符。

6、 re.X或re.VERBOSE: 通过该旗标允许分行书写正则表达式，也允许为正则表达式添加注释。

In [33]:
re.findall(r'abc', 'Abc is a good domain, ABC is good', flags=re.I)

['Abc', 'ABC']

In [34]:
# 以下两个正则表达式都可以匹配上海市的座机号
a = re.compile(r"""021    # 上海的区号
                \-        # 中间的短横线
                \d{8}     # 8个数值""", flags=re.X)
                
b = re.compile(r'020\-\d{8}')

### 正则表达式中支持的特殊字符
- $ : 匹配一行的结尾
- ^ : 匹配一行的开头
- \(\) : 标记子表达式，也就是组的开始位置和结束位置
- \[\] : 用于确定中括号表达式的开始位置和结束位置
- \{\} : 用于标记前面子表达式出现的频度
- \* : 指定前面子表达式可以出现零次或多次
- \+ : 指定前面子表达式可以出现一次或多次
- ? : 指定前面子表达式可以出现零次或一次
- . : 匹配除换行符之外的任意单个字符
- | : 指定在两项之间任选一项  

### 正则表达式支持的预定义字符（通配符）
- \d : 匹配0-9所有数字
- \D : 匹配非数字
- \s : 匹配所有的空白字符，包括空格、制表符、回车符、换页符、换行符
- \S : 匹配所有的非空白字符
- \w : 匹配所有的单词字符，包括0-9的所有数字、26个英文字母和下划线
- \W : 匹配所有的非单词字符  

### 方括号表达式
- 表示枚举：例如\[abc\]表示匹配a, b, c中任意一个字符
- 表示范围：例如\[a-f\]表示匹配a-f范围内的任意字符
- 表示求否^：例如\[^abc\]匹配非a, b, c的任意字符; \[^a-f\]表示不是a-f范围内的任意字符

### 容器相关类 --> set和frozenset  
set集合有两个主要特征：1. set不记录元素的添加顺序；2. 元素不允许重复。但set集合是可变容器，程序可以改变容器中的元素。与set对应的还有frozenset，frozenset是set的不可变版本，它的元素是不可变的。

In [35]:
# 使用{}构建set集合
c = {'kk'}
c.add('master')
c.add(6)
c

{6, 'kk', 'master'}

In [36]:
# 使用set函数创建set集合
books = set()
books.add('Java')
books.add('Go')
books

{'Go', 'Java'}

In [37]:
books.clear()
books

set()

1、 issubset(): 判断前面的set是否为后面的set集合的子集合

2、 issuperset(): 判断前面的set是否为后面的set集合的父集合

3、 difference(): 用前面的set的元素减去后面的set集合的元素，返回两个集合做减法的结果

4、 difference_update(): 改变第一个set集合

5、 intersection(): 返回两个set集合的交集

6、 intersection_update(): 通过交集运算改变第一个集合

7、 union(): 返回两个set的并集

8、 update(): 通过并集运算改变第一个set

In [38]:
s = set()
f_s = frozenset('Kotlin')
# 为set添加frozenset
s.add(f_s)
s

{frozenset({'K', 'i', 'l', 'n', 'o', 't'})}

In [39]:
sub_s = {'Java'}
# 为set添加set
try:
    s.add(sub_s)
except TypeError:
    print("unhashable type: 'set'")

unhashable type: 'set'


### 容器相关类 --> 双端队列(deque)
双端队列代表一种特殊的队列，它可以在两端同时进行插入、删除操作。对于一个deque来说，它可以从两端分别进行插入、删除操作，如果程序将所有的插入、删除操作都固定在一端进行，那么这个deque就变成了栈；如果固定在一端只添加元素，在另一端只删除元素，那么它就变成了队列。

In [40]:
from collections import deque

a = [e for e in dir(deque) if not e.startswith('_')]
print(a)

['append', 'appendleft', 'clear', 'copy', 'count', 'extend', 'extendleft', 'index', 'insert', 'maxlen', 'pop', 'popleft', 'remove', 'reverse', 'rotate']


In [41]:
stack = deque(('Java', 'Python'))

# 元素入栈
stack.append('Go')
stack.append('Scala')
stack

deque(['Java', 'Python', 'Go', 'Scala'])

In [42]:
# 元素出栈，后添加的元素先出栈
stack.pop()

'Scala'

In [43]:
stack.popleft()

'Java'

deque还有一个rotate()方法，该方法可将队列的队尾元素移动到队头，使之首尾相连:

In [44]:
q = deque(range(5))
q

deque([0, 1, 2, 3, 4])

In [45]:
q.rotate()
q

deque([4, 0, 1, 2, 3])

### Python的堆操作
堆是一种经过排序的完全二叉树，若树中所有节点的值都小于其左、右子节点的值，次数的根节点值最小，这种称为小顶堆，反之称为大顶堆。小顶堆的任意子树也是小顶堆。Python提供基于小顶堆的操作，Python并没有提供“堆”这种数据类型，它是直接把list当成堆处理，这样程序每次获取堆元素时，总会取得堆中最小的元素:

In [46]:
import heapq

print(heapq.__all__)

['heappush', 'heappop', 'heapify', 'heapreplace', 'merge', 'nlargest', 'nsmallest', 'heappushpop']


In [47]:
test_list = list(range(10))
test_list.append(0.5)
test_list

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0.5]

In [48]:
# 将堆属性应用到列表上
heapq.heapify(test_list)
# 满足小顶堆的特征
test_list

[0, 0.5, 2, 3, 1, 5, 6, 7, 8, 9, 4]

In [49]:
# heappush(heap, item): 将item元素加入堆
heapq.heappush(test_list, 7.2)
test_list

[0, 0.5, 2, 3, 1, 5, 6, 7, 8, 9, 4, 7.2]

In [50]:
# 将堆中最小元素弹出
heapq.heappop(test_list)

0

In [51]:
heapq.heappop(test_list)

0.5

In [52]:
# 将堆中最小元素弹出，并将x元素入堆
heapq.heapreplace(test_list, 8.1)
test_list

[2, 3, 5, 7, 4, 8.1, 6, 7.2, 8, 9]

In [53]:
# 将item入堆，然后弹出堆中最小元素
heapq.heappushpop(test_list, 3.4)

2

In [54]:
# 返回堆中最大的n个元素
heapq.nlargest(4, test_list)

[9, 8.1, 8, 7.2]

In [55]:
# 返回堆中最小的n个元素
heapq.nsmallest(4, test_list)

[3, 3.4, 4, 5]

当程序要获取列表中最大的n个元素或者最小的n个元素时，使用堆能缓存列表的排序结果，因此具有较好的性能。

### 容器相关类 --> ChainMap对象
ChainMap使用链的方式将多个dict“链”在一起，从而允许程序可以直接获取任意一个dict所包含的key对应的value。但实际上其底层并未真正合并这些dict，因此程序无须调用多个update()方法将多个dict进行合并。由于并未真正合并，所以在多个dict中完全可能具有重复的key，此时排在“链”前面的dict中的key具有更高的优先级。

In [56]:
from collections import ChainMap

a = {'Java': 88, 'Python': 96}
b = {'Go': 83, 'Python': 92}
c = {'Swift': 85, 'Go': 81}
cm = ChainMap(a, b, c)
cm

ChainMap({'Java': 88, 'Python': 96}, {'Go': 83, 'Python': 92}, {'Swift': 85, 'Go': 81})

In [57]:
cm['Java']

88

借鉴Python文档库中的一个例子，该例子将局部范围的定义、全局范围的定义、Python内置定义链成一个ChainMap，当程序通过该ChainMap获取变量时，将会按照局部定义、全局定义、内置定义的顺序执行搜索：

In [58]:
import builtins

my_name = 'leon'
def test():
    my_name = 'faye'
    pylookup = ChainMap(locals(), globals(), vars(builtins))
    # 访问my_name对应的value，优先使用局部范围的定义
    print(pylookup['my_name'])
    # 访问len对应的value，由于在局部范围、全局范围中都找不到，因此访问内置的len函数
    print(pylookup['len'])

In [59]:
test()

faye
<built-in function len>


### 容器相关类 --> OrderedDict对象
该对象是dict的子类，其最大特征是：它可以维护添加key-value对的顺序，即先添加的key-value排在前面，后添加的key-value排在后面。即使两个OrderedDict中的键值对完全相同，但顺序不同，程序在判断它们是否相等的时候也依然会返回False:

In [60]:
from collections import OrderedDict

od = OrderedDict(b = 5, c = 2, a = 7)
od

OrderedDict([('b', 5), ('c', 2), ('a', 7)])

In [61]:
d = OrderedDict()
d['python'] = 98
d['java'] = 94
d['go'] = 82
d['kotlin'] = 85

for k, v in d.items():
    print(k, v)

python 98
java 94
go 82
kotlin 85


由于OrderedDict是有序的，因此Python为之提供了如下两个方法：

1、 popitem(last=True): 默认弹出并返回最右边（最后加入）的key-value对；如果将last参数设置为False，则弹出并返回最左边的key-value对

2、 move_to_end(key, last=True): 默认将指定的key-value对移动到最右边（最后加入）；如果将last设置为False，则移到最左边

### 函数相关模块 --> itertools
首先是itertools模块中三个生成无限迭代器的函数：

1、count(start, \[step\]): 生成start, start+step, start+2*step ...的迭代器，step默认为1

2、cycle(p): 对序列p生成无限循环的迭代器，例如cycle('abcd')生成的迭代器为：a,b,c,d,a,b,c,d,......

3、repeat(elem, \[n\]): 生成无限个elem元素重复的迭代器

In [62]:
import itertools

for e in itertools.count(10, 3):
    print(e, end=' ')
    if e > 20:
        break

10 13 16 19 22 

In [63]:
my_counter = 0
for e in itertools.cycle(['python', 'java', 'go']):
    print(e, end=' ')
    my_counter += 1
    if my_counter > 7:
        break

python java go python java go python java 

In [64]:
for e in itertools.repeat('go', 3):
    print(e, end='\t')

go	go	go	

在itertools模块中还有一些常用的迭代器函数:

In [65]:
# 使用累加的方式计算下一个元素的值
for e in itertools.accumulate(range(6)):
    print(e, end=' ')

0 1 3 6 10 15 

In [66]:
# 使用x*y的方式来计算下一个元素的值
for e in itertools.accumulate(range(1, 6), lambda x, y: x * y):
    print(e, end=' ')

1 2 6 24 120 

In [67]:
# 将两个序列链在一起，生成新的迭代器
for e in itertools.chain(['a', 'b'], ['go', 'python']):
    print(e, end=' ')

a b go python 

In [68]:
# 根据第二个序列来筛选第一个序列的元素
# 由于第二个元素只有中间两个元素为1（True），因此第一个序列只保留中间两个元素
for e in itertools.compress(['a', 'b', 'python', 'go'], [0, 1, 1, 0]):
    print(e, end=' ')

b python 

In [69]:
# 获取序列中长度不小于4的元素开始到结束的所有元素
for e in itertools.dropwhile(lambda x: len(x)<4, ['a', 'b', 'kotlin', 'x', 'y']):
    print(e, end=' ')

kotlin x y 

In [70]:
# 去掉序列中从长度不小于4的元素开始到结束的所有元素
for e in itertools.takewhile(lambda x: len(x)<4, ['a', 'b', 'kotlin', 'x', 'y']):
    print(e, end=' ')

a b 

In [71]:
# 只保留序列中长度不小于4的元素
for e in itertools.filterfalse(lambda x: len(x)<4, ['a', 'b', 'kotlin', 'x', 'y']):
    print(e, end=' ')

kotlin 

In [72]:
# 使用pow函数对原序列的元素进行计算，将计算结果作为新序列的元素
for e in itertools.starmap(pow, [(2, 5), (3, 2), (10, 3)]):
    print(e, end=' ')

32 9 1000 

In [73]:
# 将'ABCD'、'xy'的元素按索引合并成元组，这些元组作为新序列的元素
# 长度不够的元素使用'-'字符代替
for e in itertools.zip_longest('ABCD', 'xy', fillvalue='-'):
    print(e, end=' ')

('A', 'x') ('B', 'y') ('C', '-') ('D', '-') 

在itertools中还有一些用于生成排列组合的工具函数:

In [74]:
# 使用两个序列进行排列组合
for e in itertools.product('AB', 'CD'):
    print(e, end=' ')

('A', 'C') ('A', 'D') ('B', 'C') ('B', 'D') 

In [75]:
# 使用一个序列，重复两次进行全排列
for e in itertools.product('AB', repeat=2):
    print(e, end=' ')

('A', 'A') ('A', 'B') ('B', 'A') ('B', 'B') 

In [76]:
# 从序列中取两个元素进行排列
for e in itertools.permutations('ABCD', 2):
    print('.'.join(e), end=', ')

A.B, A.C, A.D, B.A, B.C, B.D, C.A, C.B, C.D, D.A, D.B, D.C, 

In [77]:
# 从序列中取两个元素进行组合，元素不允许重复
for e in itertools.combinations('ABCD', 2):
    print('.'.join(e), end=', ')

A.B, A.C, A.D, B.C, B.D, C.D, 

In [78]:
# 从序列中取两个元素进行组合，元素允许重复
for e in itertools.combinations_with_replacement('ABCD', 2):
    print('.'.join(e), end=', ')

A.A, A.B, A.C, A.D, B.B, B.C, B.D, C.C, C.D, D.D, 

### 函数相关模块 --> functools的功能函数

In [79]:
import functools

a = [e for e in dir(functools) if not e.startswith('_')]
print(a)

['RLock', 'WRAPPER_ASSIGNMENTS', 'WRAPPER_UPDATES', 'cached_property', 'cmp_to_key', 'get_cache_token', 'lru_cache', 'namedtuple', 'partial', 'partialmethod', 'recursive_repr', 'reduce', 'singledispatch', 'singledispatchmethod', 'total_ordering', 'update_wrapper', 'wraps']


### decimal
在重视精度的场合使用decimal，例如货币计算:

In [80]:
rate = 1.45
seconds = 3 * 60 + 42
cost = rate * seconds / 60
print(cost)
print(round(cost, 2))

5.364999999999999
5.36


如果希望round是把数字调高至5.37。为了解决这种问题，可以使用内置的Decimal类，该类默认提供28个小数位，以进行定点数学运算。Decimal类解决了IEEE 754浮点数产生的精度问题，而且开发者可以更加精准地控制该类的舍入行为:

In [81]:
from decimal import *

rate = Decimal('1.45')
seconds = Decimal('222')
cost = rate * seconds / Decimal('60')
rounded = cost.quantize(Decimal('0.01'), rounding=ROUND_UP)
print(cost)
print(rounded)

5.365
5.37
